KAFKA-12313: Streamling windowed Deserialiser configs.

vamossagar12 opened the pull request:
This PR aims to streamline the configurations for WindowedDeserialisers. It deprecates default.windowed.key.serde.inner and default.windowed.value.serde.inner configs in StreamConfig and adds window.inner.class.deserialiser. Details described here: https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=177047930

vamossagar12 committed:
KAFKA-12313: KIP-725: Streamlining configs for Windowed Deserialisers

vamossagar12 commented:
hey @ableegoldman , plz review whenever you get the chance..

ableegoldman commented on streams/examples/src/main/java/org/apache/kafka/streams/examples/pageview/PageViewTypedDemo.java:
```diff
@@ -177,9 +177,8 @@ public static void main(final String[] args) {
         props.put(StreamsConfig.DEFAULT_TIMESTAMP_EXTRACTOR_CLASS_CONFIG, JsonTimestampExtractor.class);
         props.put(StreamsConfig.DEFAULT_KEY_SERDE_CLASS_CONFIG, JSONSerde.class);
-        props.put(StreamsConfig.DEFAULT_WINDOWED_KEY_SERDE_INNER_CLASS, JSONSerde.class);
+        props.put(StreamsConfig.WINDOW_INNER_CLASS_DESERIALISER, JSONSerde.class.getName());
```
We should be providing an instance of the TimeWindowedSerde to the Streams DSL in the application topology, not using this config. That's the idea of this KIP -- a Kafka Streams application should pass in a `new TimeWindowedSerde(innerClass, windowSize)` rather than using this like a default.

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
```diff
@@ -380,16 +380,23 @@
     private static final String DEFAULT_WINDOWED_VALUE_SERDE_INNER_CLASS_DOC = "Default serializer / deserializer for the inner class of a windowed value. Must implement the " +
         "<code>org.apache.kafka.common.serialization.Serde</code> interface.";
 
+    public static final String WINDOW_INNER_CLASS_DESERIALISER = "window.inner.class.deserializer";
```
I know this is the name we voted on in the KIP, but I wonder if it should be `WINDOWED` instead of `WINDOW`? I honestly didn't notice it before, but "windowed" seems more consistent with the actual Serde name. 
If you agree, you can just update the KIP and send a quick note to the KIP thread on the mailing list to clarify the name change

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
```diff
@@ -648,16 +655,11 @@
+            .define(WINDOW_INNER_CLASS_DESERIALISER,
                     Type.CLASS,
                     null,
                     Importance.MEDIUM,
```
This should be `LOW` since the corresponding config for window size of a windowed serde is also `LOW`. The configs are grouped by Importance so you'll need to move it to the section below (and also it should be in alphabetical order).

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/kstream/SessionWindowedDeserializer.java:
```diff
@@ -45,16 +38,26 @@ public SessionWindowedDeserializer(final Deserializer<T> inner) {
+
+        if (inner != null && windowInnerClassDeserialiserConfig != null) {
+            if (!inner.getClass().getName().equals(windowInnerClassDeserialiserConfig))
+                throw new IllegalArgumentException("Inner class deserializer config set using time windowed " +
```
We should log an error that prints out what the two configs actually are

ableegoldman reviewed (commented):
Hm... I may have said something misleading during the KIP discussion, which I only just realized from seeing this. In the case of the window.size config that we added in KIP-659, we only cared about adding the config for the console consumer, since the console producer does not encode the window size in the serialized data. I think I said the same was true for the inner class serde but actually that's not true, both the consumer and producer need to know this inner serde to actually de/serialize the data. 

So, I think we actually do want the new config to be `WINDOW_INNER_CLASS_SERDE` rather than just `WINDOW_INNER_CLASS_SERIALIZER`, and continue to allow it to be used to configure for example the SessionWindowedSerializer.

Sorry for the confusion 😅  -- seems like everyone else overlooked this during the KIP discussion so I don't think this correction will be too controversial, or require a new vote. You can just update the KIP doc and send a quick note to the DISCUSS thread to announce the config name change.

Also, I wonder if it should be WINDOWED instead of WINDOW? I honestly didn't notice the lack of the "ed" in the config name before, but "windowed" seems more consistent with the actual Serde name. If you agree, let's include that in the KIP update as well. ie

`WINDOW_INNER_CLASS_SERIALIZER` --> `WINDOWED_INNER_CLASS_SERDE`

vamossagar12 commented:
Thanks @ableegoldman , actually i think i had updated the ticket as well that we need to include serialisers as well. i have updated the KIP. and sent it again for discussion. Thanks!

ableegoldman commented:
Hey @vamossagar12 , I think we're good to go on the update you proposed to this KIP. Take a look at the other feedback I left and just ping me when the PR is ready for review again. Thanks!

vamossagar12 commented:
@ableegoldman , i have addressed all the comments.. plz review whenever you get the chance.

vamossagar12 commented:
hey @ableegoldman , could you plz review the PR whenever you get the chance? Thanks!

ableegoldman reviewed (approved):
LGTM! Thanks for taking up this work and driving the KIP.

Looks like just the usual flaky `RaftClusterTest` failures, I will merge to trunk

ableegoldman commented:
Done! 👏  

@vamossagar12 I'll close out the JIRA ticket, but can you update the KIP and move it to the "Adopted" section in both the main KIP page and the Streams KIPs subpage? I can send you a link if you're not sure what I'm referring to here.

vamossagar12 commented:
thanks @ableegoldman , have updated the main KIP page, the streams kiP sub page and the actual kiP. Thanks for your support!
